Copy Java attacher jar to a tmp directory

eyalkoren opened the pull request:
## Motivation/summary

Fixing a bug discovered while testing this end-to-end: when Elastic Agent is ran through `./elastic-agent install`, it creates a file tree that is not readable for all users on Linux by default. 
Therefore, when running the attacher as a non-root user, it can't access the attacher jar.
The fix is to create a tmp directory with write permissions only to root and copy the attacher jar to it with `0755` permissions to make it accessible for all users for read and execute.

Fixe verified on:
- Amazon Linux 2
- Ubuntu 22.04

Note for reviewer:
- pay extra care to security concerns I may have overlooked
- currently, the agent is downloaded and verified for each attachment, but we should consider ways to prevent that
- we currently assume the default temp dir would work, but in some OSs we may need to have a tmp dir per user
- the new test may fail on Windows, where file modes are handled differently, we'll take care of that later

## Checklist

- [x] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Documentation has been updated - would be handled separately through #8708 
- [x] Added unit tests

## How to manually test these changes

1. setup a stack and install a matching agent on the tested Linux host
2. set Elastic agent log level to `debug`
3. run a Java app on the tested host **as a limited-privileges user.** It is recommended that this app will have a way to identify it specifically through matching rules, for example, use `-Delastic.apm.attach=true` in the command line that runs it
4. within the Fleet UI, edit the corresponding agent's policy to include the APM integration. In the APM integration, within the APM agent tab, enable the Java agent auto-attachment option and make sure the matching rule includes a hint that will specifically identify your Java app process
5. apply the policy and inspect the logs. It should contain indications that attachment took place. In addition, it should contain the path of the temp dir to which the attacher was copied
6. login to the Linux host and look for the tmp dir indicated by logs - it should have `0700` mode and be owned by the same user that runs the Java app. It should contain the attacher jar with `0600` mode and be owned by the same user. If attachment was successful, it should contain an `agent` folder with internal structure that includes some PGP-utility libraries and a directory per agent version as requested within the integration.
7. verify that APM data is being traced and available in Kibana
8. restart the Java app, invoke some traffic on it and verify that the same agent jar was used and that APM data is available 
9. stop Elastic agent and verify that the attacher temp dir is removed

## Related issues

#8718 
#8660 
#6969 
#7832

apmmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-8803/detail/PR-8803/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-8803/detail/PR-8803/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-8803/detail/PR-8803/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-8803/detail/PR-8803/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-server_8803.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/apm-ci/transactions/view?rangeFrom=2022-08-04T15:41:26.753Z&rangeTo=2022-08-04T16:01:26.753Z&transactionName=BUILD+apm-server%2Fapm-server-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=132b803f6e1096c15e5d35e9dd91f904&transactionId=4cce2143ca599d2e)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-08-04T15:51:26.753+0000


* Duration: 25 min 16 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 137  |
| Skipped      | 1 |
| Total        | 138   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate and publish the docker images.


- `/test windows` : Build & tests on Windows.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

apmmachine commented:
### :books: Go benchmark report

Diff with the `main` branch

```
name                                                            old time/op    new time/op    delta
pkg:github.com/elastic/apm-server/internal/agentcfg goos:linux goarch:amd64
FetchAndAdd/FetchFromCache-12                                     41.9ns ± 0%    42.7ns ± 3%    +1.89%  (p=0.016 n=4+5)
FetchAndAdd/FetchAndAddToCache-12                                  100ns ± 3%     105ns ± 3%    +4.73%  (p=0.008 n=5+5)
pkg:github.com/elastic/apm-server/internal/decoder goos:linux goarch:amd64
CompressedRequestReader/deflate_sniff-12                          17.2µs ± 4%    10.5µs ±65%   -39.27%  (p=0.016 n=5+5)
pkg:github.com/elastic/apm-server/internal/model/modelindexer goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/processor/stream goos:linux goarch:amd64
BackendProcessor-12                                               0.00ns ±24%    0.00ns ±54%  +301.19%  (p=0.016 n=4+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel8-12    0.00ns ±26%    0.00ns ±59%   +92.28%  (p=0.016 n=5+5)
pkg:github.com/elastic/apm-server/internal/publish goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:linux goarch:amd64
AggregateTransaction-12                                           72.2ns ± 0%    72.9ns ± 1%    +0.99%  (p=0.032 n=5+5)
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:linux goarch:amd64
ReadEvents/json_codec/399_events-12                               2.63ms ± 1%    2.66ms ± 1%    +1.10%  (p=0.016 n=5+5)
ReadEvents/nop_codec_big_tx/0_events-12                            312ns ± 6%     324ns ± 3%    +3.85%  (p=0.032 n=5+5)
IsTraceSampled/unknown-12                                          356ns ± 2%     371ns ± 3%    +3.96%  (p=0.032 n=5+5)

name                                                            old alloc/op   new alloc/op   delta
pkg:github.com/elastic/apm-server/internal/agentcfg goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/decoder goos:linux goarch:amd64
CompressedRequestReader/gzip_sniff-12                             45.7kB ± 0%    45.6kB ± 0%    -0.21%  (p=0.032 n=5+5)
CompressedRequestReader/deflate_sniff-12                          45.1kB ± 0%    44.9kB ± 0%    -0.32%  (p=0.016 n=5+4)
pkg:github.com/elastic/apm-server/internal/model/modelindexer goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/processor/stream goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/publish goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:linux goarch:amd64
WriteTransaction/json_codec_big_tx-12                             3.52kB ± 0%    3.52kB ± 0%    +0.03%  (p=0.016 n=5+4)
ReadEvents/json_codec/1000_events-12                              3.20MB ± 1%    3.23MB ± 0%    +0.71%  (p=0.048 n=5+4)
ReadEvents/nop_codec/1_events-12                                  3.27kB ± 0%    3.27kB ± 0%    +0.09%  (p=0.024 n=5+5)
ReadEvents/nop_codec_big_tx/100_events-12                          250kB ± 0%     251kB ± 0%    +0.26%  (p=0.008 n=5+5)

name                                                            old allocs/op  new allocs/op  delta
pkg:github.com/elastic/apm-server/internal/agentcfg goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/decoder goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/model/modelindexer goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/processor/stream goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/publish goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:linux goarch:amd64

```

_report generated with https://pkg.go.dev/golang.org/x/perf/cmd/benchstat_
<!--COMMENT_GENERATED_WITH_ID_bench.diff-->

eyalkoren commented:
/test windows

eyalkoren commented on internal/beater/java_attacher/java_attacher.go:
```diff
@@ -145,7 +149,13 @@ func (j *JavaAttacher) Run(ctx context.Context) error {
 	}
 
-	// Non-Windows: run discovery and attachment until context is closed.
+	tempDir, err := j.createAttacherTempDir()
+	if err != nil {
+		j.logger.Errorf("failed to create a Java attacher temp dir: %v. Using the bundled attacher jar instead", err)
+	} else {
+		defer j.deleteTempDir(tempDir)
+	}
```
I didn't bother looking into all possible errors returned by everything I used to create the temp dir, however, if we default to using the bundled attacher, it will still work if the application user is the same as the Elastic Agent user. I agree this is not likely in production, but could be a useful fallback for users that just want to explore on some test env or even locally on their laptop.

axw replied:
> but could be a useful fallback for users that just want to explore on some test env or even locally on their laptop.

OK. This is not a good enough reason to increase the code complexity, so please just return the error. Failure to create a temporary file suggests a more serious issue, like the disk being full.

eyalkoren replied:
OK, will do.
Explaining the rationale- the way I see it, ad-hoc trial is a very important use case and an error here can be anything, for example such raised by the user's laptop OS implementation indicating it does not support it.

eyalkoren replied:
Still needs changing, will change that shortly

eyalkoren replied:
Currently, we still fallback to the bundled attacher jar.
Pro: we may be able to attach to a specific user but not to other
Con: we'll try every time we see a JVM for that user (but only once per JVM)

LMK if you want something else, for example:
- bail and exit the attacher upon the first failure
- cache failure per user using a special value (so that we can still attach to other users' JVMs)

axw replied:
Please log & cache the failure, and continue attempting for other users.

eyalkoren commented on internal/beater/java_attacher/java_attacher.go:
```diff
@@ -361,8 +436,12 @@ func (j *JavaAttacher) attachJVM(ctx context.Context, jvm *jvmDetails) error {
 }
 
 func (j *JavaAttacher) attachJVMCommand(ctx context.Context, jvm *jvmDetails) *exec.Cmd {
+	attacherJar := bundledJavaAttacher
+	if j.tempAttacherDir != "" {
+		attacherJar = tmpJavaAttacherJarPath(j.tempAttacherDir)
+	}
```
Yes, I don't like it either, but since this is called from `attachJvm`, which is used within `foreachJVM` that currently requires a specific function structure, I did it this way. I could change that of course, but this is planned to change soon to support a tmp dir per user, where it will make more sense that the more complex state (a map from user to tmp-dir) is maintained by the attacher (you cannot do that from `Run` if you loop over multiple JVMs). 

> Instead of storing tempAttacherDir, we could store a path to the attacher jar, whether it's in a temp directory or not.

But I want to delete the tmp dir and its contents when `Run` exits, so I want a reference to that

eyalkoren replied:
Changed to tmp dir per user as planned

eyalkoren commented on internal/beater/java_attacher/java_attacher.go:
```diff
@@ -167,6 +177,71 @@ func (j *JavaAttacher) Run(ctx context.Context) error {
+func (j *JavaAttacher) closeFile(file *os.File) {
+	err := file.Close()
+	if err != nil {
+		j.logger.Errorf("failed to close file %v: %v", file, err)
```
The way I see it, I don't need to bother thinking about all possible errors and I assume they shouldn't occur, but if they do, for whatever reason I couldn't think of, I'd rather have it logged as it may indicate I am doing something not as I planned. Wherever I don't log something, it is because I am lazy 🙂 
But I definitely don't mind ignoring those and simplifying the code.

axw replied:
Yeah I understand the rationale. I would sometimes do this too, but in this case I don't think it's justified. Failure to close the file means something more generally catastrophic has happened, and we'll find out soon enough from another avenue. Let's just do the defer for now.

eyalkoren replied:
Oops, missed that, changing shortly

eyalkoren commented on internal/beater/java_attacher/java_attacher.go:
```diff
@@ -167,6 +177,71 @@ func (j *JavaAttacher) Run(ctx context.Context) error {
+	if err != nil {
+		return "", err
+	}
+	// todo: check if we need to use different value for specific OSs
```
This comment was written because of the referred documentation. For example, once we support Windows, we may want the per-user attacher jar to have the `0200` bit off, which is not the case today.

eyalkoren replied:
Irrelevant in light of what we now know about running on Windows (it's possible to attach to any JVM as the administrator)

eyalkoren commented on internal/beater/java_attacher/java_attacher.go:
```diff
@@ -167,6 +177,71 @@ func (j *JavaAttacher) Run(ctx context.Context) error {
+	// make sure this dir is readable and executable by all users and writable only by the current user
+	err = os.Chmod(tempDir, 0755)
+	if err != nil {
+		j.logger.Errorf("failed to change permissions for %v to make it readable for all users: %v", tempDir, err)
```
IIUC, the default permissions of the created tmp dir are restrictive and here I am only try to make them more permissive. Like explained [above](https://github.com/elastic/apm-server/pull/8803#discussion_r940976581), I prefer a fallback to only work when the app user is the same as the Elastic Agent user than never work.
If you are still not convinced lmk and I will change accordingly (fail on any error)

eyalkoren replied:
BTW, @jackshirazi and I just tested on Windows and saw that the fallback WOULD indeed work on Windows, where I suppose `Chmod` is more likely to return an error.

eyalkoren replied:
Non tolerant now to any failure in tmp dir creation

eyalkoren commented on internal/beater/java_attacher/java_attacher.go:
```diff
@@ -167,6 +177,71 @@ func (j *JavaAttacher) Run(ctx context.Context) error {
+		return "", fmt.Errorf("failed to copy bundled attacher jar to %v: %w", bundledJavaAttacher, err)
+	}
+	j.logger.Debugf("%v (%v bytes) successfully copied to %v", bundledJavaAttacher, nBytes, tmpAttacherJarPath)
+	err = os.Chmod(tmpAttacherJarPath, 0755)
```
I thought so to, but luckily I tested and apparently it does in some cases.

axw replied:
Can you please elaborate? What are those cases?

eyalkoren replied:
Amazon Linux, at least

axw replied:
Please elaborate further: what happens when you use mode `0644` for the attacher jar on that platform?

eyalkoren replied:
It failed with an access failure message. Trying with `0744` failed and with `0755` succeeded

axw replied:
I'm trying to understand what is causing the failure, so please give all the details. What is the error message? Maybe if you can provide the steps you took to test it, I can see it for myself.

I'm having a hard time believing that the file needs to be executable. Did you perhaps set both the temp directory _and_ the file to 0644? If that were the case, then it would make sense. Directories do need to be executable.

eyalkoren replied:
> Maybe if you can provide the steps you took to test it, I can see it for myself

I built the APM Server, replaced the binaries on my Linux machine within my Elastic Agent installation and restarted the agent, then restarted the test app. 

> Did you perhaps set both the temp directory and the file to 0644? 

Very likely! IIRC, I tried only when both were either `0744` or `0755`

>  Directories do need to be executable

I didn't know that! Makes even less sense to me than requiring the jar to executable, given that the actual executable is the java binary...

I will definitely try it out. I would have reproduced now, but the current version is 8.5-SNAPSHOT while I am testing with 8.4-SNAPSHOT and merging with 8.4 requires some effort. I'd be happy to know of an easy way to build as a different version. I will get to testing that later today, preferably after we discuss offline.

eyalkoren replied:
I thought the comment above was specific to running `java` and pointing to the target jar, but it's about the directory that needs to be executable in Linux in order to be accessible in general, thanks for pointing it out @axw 🙏

eyalkoren replied:
Modes are `0700` for the tmp dir and `0600` for the jar, and the owner is set to the application user

eyalkoren commented on internal/beater/java_attacher/java_attacher.go:
```diff
@@ -167,6 +177,71 @@ func (j *JavaAttacher) Run(ctx context.Context) error {
+// maybe by externally executing `echo XXX` and reading the output, relying on environment
+// variables as implemented in os.TempDir, or looking in known paths.
+// In such case, we would switch to using a cache of temp-agent-dir per user.
+func (j *JavaAttacher) createAttacherTempDir() (string, error) {
```
I don't want exclusive access, the attacher is ran in a separate process by another user. In addition, as said above, the plan is to generically manage temp-dir per user with proper write permissions, so it will be a similar structure anyway

eyalkoren replied:
See below

eyalkoren replied:
This was necessary so that we can cache agent jars instead of downloading every time, and in order to be consistent with what we do in Windows. However, since we learned that we may not be required to have a tmp dir per user in Windows, I will apply this suggestion and we will live without the performance gain of caching agent jars

eyalkoren replied:
Implemented tmp dir per user

eyalkoren commented on internal/beater/java_attacher/java_attacher.go:
```diff
@@ -167,6 +177,71 @@ func (j *JavaAttacher) Run(ctx context.Context) error {
+// maybe by externally executing `echo XXX` and reading the output, relying on environment
+// variables as implemented in os.TempDir, or looking in known paths.
+// In such case, we would switch to using a cache of temp-agent-dir per user.
+func (j *JavaAttacher) createAttacherTempDir() (string, error) {
```
I am planning to change that very soon, so that there will be a directory per user with the proper permissions, including write. This way, in combination with `os.Chown`, we can cache the attacher jar instead of downloading over and over for any attachment.

axw replied:
Let me see if I understand correctly:
 - the bundled attacher will download the full attacher from Maven, and we would like that to be cached
 - this needs to be cached per user (because it is downloaded by a process running as that user)

Does that mean the bundled attacher needs to be copied for each user though? Does the directory holding the bundled attacher and the directory in which the attacher is downloaded have to be the same?

eyalkoren replied:
I will change as proposed and we can rethink if we want to introduce agent jar caching as described [above](https://github.com/elastic/apm-server/pull/8803#discussion_r942079198).

For agent jar caching, we need the attacher process to run as the target (application) user, and make sure the current directory is only writable to the current user. The way to ensure that is create a tmp dir per user, make this user the owner of that dir and set its mode to `0700`.

eyalkoren replied:
Implemented tmp dir per user

axw commented on internal/beater/java_attacher/java_attacher.go:
```diff
@@ -167,6 +177,71 @@ func (j *JavaAttacher) Run(ctx context.Context) error {
+}
+
+// Creates a temp dir for the java attacher jar with the attacher jar in it, both with 0755 access permissions.
+// The assumption is that the default temp dir will be readable for all users.
+// However, there may be systems in which the temp directory is user-specific,
+// in which case we'll need to modify this function to run a temp dir search per user,
+// maybe by externally executing `echo XXX` and reading the output, relying on environment
+// variables as implemented in os.TempDir, or looking in known paths.
+// In such case, we would switch to using a cache of temp-agent-dir per user.
```
SGTM, thanks

eyalkoren replied:
removed

eyalkoren commented:
/test windows

eyalkoren commented:
Status update: 

- Windows: it appears that on Windows, a JVM that runs as administrator can runtime attach to a JVM that is ran by another user. Therefore, copying to a temp folder may not even be required and the bundled attacher can be used directly.

- Linux: if we copy the attacher jar directly to the root temp dir, it would be accessible to all users with `0644` mode. Once a JVM is matched against an include rule, the attacher is invoked in the context of the target JVM user. It then downloads and verifies an agent jar and store it locally. Since the current directory in this context (root temp dir) is writable for that user, it will use it to cache the agent jars by creating a `<TEMP-ROOT>/agent` directory, owned by that user and with `0755` mode. Subsequent attachments by this user can use cached agents. This is safe since the `agent` folder is not writable to other users. However, if the attacher integration tries to attach to a JVM as another user, the existing cache won't be accessible for it. 

Following is how the proposed implementation, until we discover something new when testing that:
- [ ] If running on Windows - no need to copy to temp, run the embedded jar as administrator
- [ ] On Linux, keep a sub temp attacher dir for each user that we want to attach to, following these guidelines:
   - [ ] user-attacher-temp-subdir should be owned by the corresponding user and have a `0700` mode
   - [ ] make sure the directory creation and `os.Chmod` execution are done atomically
   - [ ] the attacher jar should be copied into it, be owned by the corresponding user and have a `0400` mode
   - [ ] attacher loop should be aborted upon any error with executing the above
   - [ ] at `Run` exit, all tmp dirs should be removed. This may rely on `os.RemoveAll`, but better to recursively traverse the subtree, so that if there are non-removable files (e.g. help by still running attacher process), everything else will be cleared
- [ ] macOS - needs more investigation

Placing here an offline discussion for reference- the above solution for Linux relies on us properly maintaining directory tree within a root that has a `777` access mode (the root temp dir). If we instead copy to anywhere with the user home directory, for example the `.cache` or equivalent caching folder, then all files will be inherently protected by the OS access default settings. However, we decided to got with temp based on the following considerations:
- getting the temp dir based on `os.TempDir` is generic and properly maintained
- in contrary, getting the user home dir and specific subdir (like caching dir) requires a per-OS error-prone discovery logic
- when we use `os.MkdirTemp`, the directory is created with exclusive access by the caller
- temp directory is the proper place to store temp files, such that should be removed on process termination, especially where there is risk that we fail to remove files on exit

eyalkoren commented on internal/beater/java_attacher/java_attacher.go:
```diff
@@ -40,8 +41,10 @@ import (
+const javaAttacherJarName = "java-attacher.jar"
+
+// referring the Java attacher that is bundled at the server root
+var bundledJavaAttacher = filepath.FromSlash(fmt.Sprintf("./%v", javaAttacherJarName))
```
Removed

eyalkoren commented on magefile.go:
```diff
@@ -0,0 +1,297 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
```
removed

eyalkoren commented:
Verified through manual testing on Linux. Eventual tmp directory tree is as follows:
```
[ec2-user@ip-172-31-20-154 bin]$ ls -l /tmp/
total 0
drwx------ 3 ec2-user ec2-user 44 Aug 23 08:38 elasticapmagent-114732219
drwxr-xr-x 2 ec2-user ec2-user 18 Aug 23 08:39 hsperfdata_ec2-user
[ec2-user@ip-172-31-20-154 bin]$ ls -l /tmp/elasticapmagent-114732219/
total 8340
drwxr-xr-x 3 ec2-user ec2-user      20 Aug 23 08:38 agent
-rw------- 1 ec2-user ec2-user 8538845 Aug 23 08:38 java-attacher.jar
[ec2-user@ip-172-31-20-154 bin]$ ls -l /tmp/elasticapmagent-114732219/agent/
total 0
drwxr-xr-x 3 ec2-user ec2-user 53 Aug 23 08:38 1_33_0
[ec2-user@ip-172-31-20-154 bin]$ ls -l /tmp/elasticapmagent-114732219/agent/1_33_0/
total 9408
-rw-r--r-- 1 ec2-user ec2-user 9630190 Aug 23 08:38 elastic-apm-agent-1.33.0.jar
drwxr-xr-x 2 ec2-user ec2-user      97 Aug 23 08:38 lib
```
Summary:
- tmp dir owner is app user and access mode is `0700`
- attacher jar owner is app user and access mode is `0600`
- agent downloaded and cached within tmp dir that is only accessible to app user

One comment remained unresolved, waiting for input.

In addition, I am pretty sure there are better ways to exclude tests on a specific OS than what I did and maybe there is a better way to do OS-polymorphism than this `os_utils` and `os_utils_windows` method.

eyalkoren commented on internal/beater/java_attacher/os_utils.go:
```diff
@@ -0,0 +1,146 @@
+
+func (j *JavaAttacher) cleanResources() {
+	for _, dir := range j.tmpDirs {
+		// todo delete all contents recursively in case there is still a jar held by the another process
```
👍 So IIUC, we can rely on `RemoveAll` to proceed normally with its recursion in this case, is that correct?

axw replied:
Yes, at least I'm not aware of any reason why that would be an issue.

eyalkoren commented on internal/beater/java_attacher/os_utils.go:
```diff
@@ -0,0 +1,146 @@
+		return "", fmt.Errorf("failed to copy bundled attacher jar to %v: %w", bundledJavaAttacher, err)
+	}
+	j.logger.Debugf("%v (%v bytes) successfully copied to %v", bundledJavaAttacher, nBytes, tmpAttacherJarPath)
+	err = os.Chmod(tmpAttacherJarPath, 0600)
+	if err != nil {
+		return "", fmt.Errorf("failed to change permissions of %v to 0600: %w", tmpAttacherJarPath, err)
+	}
```
Meaning - use `0644`? Makes sense, just making sure I understand where you are coming from - if we rely on the dir permissions, no other user can attach to anything with this attacher and if this user is compromised, nothing we can do anyway. Is that it?

axw replied:
No, I mean create the file with mode 0600 in the first place (see previous suggestion)

eyalkoren replied:
Huhh, missed that, thanks!

eyalkoren commented on internal/beater/java_attacher/os_utils.go:
```diff
@@ -0,0 +1,146 @@
+	if err != nil {
+		return "", fmt.Errorf("failed to change owner of %v to be %v: %w", tempDir, uidS, err)
+	}
+	err = os.Chown(tmpAttacherJarPath, int(uid), int(gid))
```
Good point about `gid` and nice convenience method I was not aware of, thanks!

eyalkoren commented on internal/beater/java_attacher/os_utils.go:
```diff
@@ -0,0 +1,146 @@
+	if err != nil {
+		return "", fmt.Errorf("failed to change permissions of %v to 0600: %w", tmpAttacherJarPath, err)
+	}
+	err = os.Chown(tempDir, int(uid), int(gid))
+	if err != nil {
+		return "", fmt.Errorf("failed to change owner of %v to be %v: %w", tempDir, uidS, err)
+	}
```
Yes, I think you are right, I noticed the same when reviewing myself, but didn't mind leaving it this way. I will change, it will simplify method calls

eyalkoren commented on internal/beater/java_attacher/java_attacher.go:
```diff
@@ -83,6 +83,8 @@ func New(cfg config.JavaAttacherConfig) (*JavaAttacher, error) {
 		downloadAgentVersion: cfg.DownloadAgentVersion,
 		rawDiscoveryRules:    cfg.DiscoveryRules,
 		jvmCache:             make(map[int]*jvmDetails),
+		uidToAttacherJar:     make(map[string]string),
+		tmpDirs:              make([]string, 0),
```
Yes, they are! 🤦 
Thanks!

marclop commented on internal/beater/java_attacher/os_utils.go:
```diff
@@ -0,0 +1,142 @@
+		j.logger.Debugf("current user: %v", currentUser)
+	}
+
+	if currentUser.Gid != jvm.gid || currentUser.Uid != jvm.uid {
+		uid, gid, err := parseUserIds(jvm.uid, jvm.gid)
+		if err != nil {
+			return err
+		}
+		cmd.SysProcAttr = &syscall.SysProcAttr{}
+		cmd.SysProcAttr.Credential = &syscall.Credential{Uid: uint32(uid), Gid: uint32(gid)}
+	}
+	return nil
```
It seems more readable to return straightaway if the users match:
```suggestion
	if currentUser.Gid == jvm.gid && currentUser.Uid == jvm.uid {
		return nil // Users match, nothing to do.
	}
	uid, gid, err := parseUserIds(jvm.uid, jvm.gid)
	if err != nil {
		return err
	}
	cmd.SysProcAttr = &syscall.SysProcAttr{
		Credential: &syscall.Credential{Uid: uint32(uid), Gid: uint32(gid)},
	}
```

eyalkoren replied:
I think for most people it does, so I should adapt 🙂 
For some reason I got used to leaning towards minimal return statements in a function/method 🤷‍♂️ 
Your field assignment definitely looks better!

marclop commented on internal/beater/java_attacher/os_utils.go:
```diff
@@ -0,0 +1,142 @@
+	if err != nil {
+		return "", fmt.Errorf("failed to open bundled attacher jar: %w", err)
+	}
+	//goland:noinspection GoUnhandledErrorResult
```
Do we need to commit IDE specific linting comments in the code?
```suggestion
```

eyalkoren replied:
Probably not!
I am used to write to a repo where all main authors use the same IDE, where it does make sense

eyalkoren commented on internal/beater/java_attacher/os_utils_windows.go:
```diff
@@ -25,3 +25,13 @@ func (j *JavaAttacher) setRunAsUser(_ *jvmDetails, _ *exec.Cmd) error {
+
+func (j *JavaAttacher) cleanResources() {
+	// nothing to do in Windows
+}
```
Sure, up to you, I relied on @simitt's proposal to combine the exclusion build instruction in the generic file with the file naming convention to restrict the build of this one only for Windows

eyalkoren replied:
Decided offline to keep as is for now

eyalkoren commented on internal/beater/java_attacher/java_attacher.go:
```diff
@@ -83,6 +84,8 @@ func New(cfg config.JavaAttacherConfig) (*JavaAttacher, error) {
 		rawDiscoveryRules:    cfg.DiscoveryRules,
 		jvmCache:             make(map[int]*jvmDetails),
+		uidToAttacherJar:     make(map[string]string),
+		tmpDirs:              make([]string, 0),
```
Right, this needs getting used to 🙂

eyalkoren commented on internal/beater/java_attacher/os_utils.go:
```diff
@@ -0,0 +1,142 @@
+	j.tmpAttacherLock.Lock()
+	defer j.tmpAttacherLock.Unlock()
+	attacherJar, uidMapped := j.uidToAttacherJar[uid]
+	if !uidMapped {
+		tmpAttacherJar, err := j.createAttacherTempDir(uid)
+		if err != nil {
+			j.logger.Errorf("failed to create tmp dir for user %v, using the bundled attacher jar", err)
+		} else {
+			attacherJar = tmpAttacherJar
+		}
+		j.uidToAttacherJar[uid] = attacherJar
+	}
+	return attacherJar
```
Yes, much better.
The original form was setting `attacherJar = ""` in case of error, but then I realized there is a zero value for strings, so I removed it an left this redundant variable...

eyalkoren commented on internal/beater/java_attacher/java_attacher_test.go:
```diff
@@ -181,3 +230,55 @@ func TestConfig(t *testing.T) {
+	// todo: what's the proper way to exclude tests for a specific OS?
+	if runtime.GOOS == "windows" {
+		return
+	}
```
wait, this is not really the same - I understand how to do something for windows, but not sure how to do something for everything-but-windows.
So let's assume I have three types of tests:
1. applicable to all OSs
2. applicable to all but windows
3. applicable to windows only

How should I construct those? For example:
1. `java_attacher_test.go` - with no build directives
2. `java_attacher_copy_test.go` with build directives to exclude windows
3. `java_attacher_copy_test_windows.go`

Would that work?

I am not sure as I understand file naming has more implications that I am aware of in the Go build workflow.

eyalkoren replied:
Probably can't use `java_attacher_copy_test_windows.go` because we need to the `_test.go` suffix so that it is actually executed

eyalkoren commented:
Update: turns out the Administrator cannot attach to any JVM after all on Windows, that was a mistake in the former tests...
So eventually we will have to figure out how to run as a different user in Windows, if it is possible at all without providing password.

Still, I'll apply the proposed tests structure changes so that we can proceed with full Linux support and Windows support only when running the agent with the same user that runs the Java app

eyalkoren commented:
I won't be available for some time to merge this.
If you feel this is close enough, please bring to completion and merge.
I will update https://github.com/elastic/apm-agent-java/issues/2196, https://github.com/elastic/apm-server/issues/8660 and https://github.com/elastic/apm-server/issues/8718 with the state of things assuming this PR is merged more-or-less as is now.

eyalkoren commented:
@elastic/apm-server - anything else still required? 
AFAIK I addressed all review comments and updated other issues in what's still required for Windows and macOS

axw commented on internal/beater/java_attacher/java_attacher_windows_test.go:
```diff
@@ -0,0 +1,57 @@
+	cfg := createTestConfig()
+	f, err := os.Create(bundledJavaAttacher)
+	require.NoError(t, err)
+	//goland:noinspection GoUnhandledErrorResult
```
please ensure all "//goland" comments are removed

axw commented on internal/beater/java_attacher/java_attacher.go:
```diff
@@ -354,15 +367,25 @@ func (j *JavaAttacher) verifyJVMExecutable(ctx context.Context, jvm *jvmDetails)
 // once the agent is attached or the context is closed; it may be blocking for long and should be called with a timeout context.
 func (j *JavaAttacher) attachJVM(ctx context.Context, jvm *jvmDetails) error {
 	cmd := j.attachJVMCommand(ctx, jvm)
+	if cmd == nil {
+		return nil
+	}
```
why would the cmd be nil?

eyalkoren replied:
It may if the attacher jar could not be created in the tmp dir.
This is how we avoid trying to copy and issuing errors again and again for all JVMs ran by a user for which we cannot attach.

axw reviewed (changes requested):
A handful of minor comments, and what looks like a bug with error handling in copying the attacher jar.

eyalkoren commented on internal/beater/java_attacher/java_attacher.go:
```diff
@@ -160,14 +165,14 @@ func (j *JavaAttacher) Run(ctx context.Context) error {
 			continue
 		}
-		if err := j.foreachJVM(ctx, jvms, j.attachJVM, true, 30*time.Second); err != nil {
+		if err := j.foreachJVM(ctx, jvms, j.attachJVM, false, 2*time.Minute); err != nil {
```
It was changed to be up-to 2 minutes, but not at the intention to increase the interval, only to allow more time before timing out. 

At the moment, the `attachJVM` method blocks until the attacher command exits or is timed out, whichever comes first. 
We may decide that this conceptually not what we want and make it completely asynchronous and not wait for either.
However, if we decide that this is conceptually what we want, the increased timeout should not affect proper attachments, only such that get blocked/hang. I increased that because my tests showed that 30 seconds is realistic even if things work as expected (this may include downloading from maven and verifying pgp signature).

axw replied:
OK. Just asking since it seems unrelated to the other changes.

eyalkoren commented on internal/beater/java_attacher/java_attacher.go:
```diff
@@ -354,15 +367,25 @@ func (j *JavaAttacher) verifyJVMExecutable(ctx context.Context, jvm *jvmDetails)
+// as well as a corresponding status change to this JavaAttacher, where the created tmp dir and jar paths are cached.
 func (j *JavaAttacher) attachJVMCommand(ctx context.Context, jvm *jvmDetails) *exec.Cmd {
+	attacherJar := j.getAttacherJar(jvm.uid)
+	if attacherJar == "" {
+		return nil
+	}
```
It may be empty if there's an error during the attempt to copy to the temp dir. I don't think there is a bug, see reply for the comment below 🙂

eyalkoren commented on internal/beater/java_attacher/os_utils.go:
```diff
@@ -0,0 +1,142 @@
+	if err != nil {
+		j.logger.Errorf("failed to create tmp dir for user %d, using the bundled attacher jar: %v", uid, err)
+	}
+	j.uidToAttacherJar[uid] = attacherJar
```
Why? We want to attempt to copy to temp only once per user ID and cache the jar we want to use for attachment every time we encounter this user. If it succeeds, we store the tmp jar path. If it fails we store an empty string to avoid keep trying every time. No reason to assume we will succeed if we failed before

axw replied:
OK, I guess the error message above is wrong then? It says:

> failed to create tmp dir for user %d, using the bundled attacher jar: %v"

eyalkoren replied:
Yes, you are right, this is a leftover and I need to change it

eyalkoren commented on internal/beater/java_attacher/os_utils.go:
```diff
@@ -0,0 +1,142 @@
+	}
+	attacherJar, err := j.createAttacherTempDir(uid)
+	if err != nil {
+		j.logger.Errorf("failed to create tmp dir for user %d, using the bundled attacher jar: %v", uid, err)
```
Well, that is a matter of point of view.
The way I see it, whenever the APM Server is configured to try to attach an agent, the APM agent is the actual focus of the integration and the APM Server is an implementation detail. The user installed an Elastic agent on a machine that runs her Java application and want to get APM traces for it. If we can't do it, it is an error. The fact that the APM Server runs locally without errors is meaningless because it has no purpose other than sending data for this Java app.
And after the multiple back and forth, we made it clear that failing to create the temp dir is not something we want to try to recover from with a fallback

axw replied:
>And after the multiple back and forth, we made it clear that failing to create the temp dir is not something we want to try to recover from with a fallback

I'm so confused. That's what I thought too, but you are trying to recover with a fallback here aren't you? You're logging an error and continuing...

eyalkoren replied:
Sorry, you are right, the message is wrong.
There is no attempt to recover for this user, we are caching the fact that we cannot attach for JVMs of this user by the special case of empty jar path. We decided before that we do not want to fail the attacher entirely in this case, because the failure can be specific to user A and may not affect user B.

axw replied:
Thanks for clarifying. I'm happy to leave it as an error log in that case, with the message amended.

eyalkoren commented on internal/beater/java_attacher/os_utils.go:
```diff
@@ -0,0 +1,142 @@
+//
+// NOTE: this method is not thread-safe, so it should not be invoked concurrently by multiple goroutines
+func (j *JavaAttacher) createAttacherTempDir(uidS string) (string, error) {
+	uid, _, err := parseUserIds(uidS, "0")
```
Yeah, I agree this is ugly, just tried to reuse the error formatting

axw commented on internal/beater/java_attacher/java_attacher.go:
```diff
@@ -354,15 +367,25 @@ func (j *JavaAttacher) verifyJVMExecutable(ctx context.Context, jvm *jvmDetails)
 func (j *JavaAttacher) attachJVMCommand(ctx context.Context, jvm *jvmDetails) *exec.Cmd {
+	attacherJar := j.getAttacherJar(jvm.uid)
+	if attacherJar == "" {
+		return nil
```
```suggestion
		// If the attacher jar is empty, this means there was an error while
		// attempting to copy the jar to a temporary directory. In this case
		// we should return nil, which will cause attachJVM to skip the process.
		return nil
```

axw reviewed (approved):
Thanks for the updates. LGTM.

eyalkoren commented:
Thanks a lot @axw and @marclop for your assistance! 🙏 
Anything else before merging? Should we add something to changelog? I am not sure, as this is still partial (Linux support mostly) and in technical preview. If so, I'd appreciate your proposal for that.

axw commented:
@eyalkoren no worries. Yes, I think a changelog entry is worthwhile. Please also update the PR description to use the [PR template](https://raw.githubusercontent.com/elastic/apm-server/main/.github/pull_request_template.md), explaining steps to be taken for manual testing. Otherwise I think it's good to merge.

eyalkoren commented:
Done, please just see that the changelog addition is done as expected.

axw commented on changelogs/head.asciidoc:
```diff
@@ -14,6 +14,8 @@ https://github.com/elastic/apm-server/compare/8.5\...main[View commits]
 [float]
 ==== Bug fixes
 - Set `message` instead of `labels.event` for Jaeger span events {pull}8765[8765]
+- Reach full support (still technical preview) for Java agent auto attachment on Linux by allowing attachment to a JVM of any user
+{pull}8803[8803]
```
```suggestion
- Reach full support (still technical preview) for Java agent auto attachment on Linux by allowing attachment to a JVM of any user {pull}8803[8803]
```

SylvainJuge commented:
Tested with 8.5.0

Fleet agent used https://github.com/elastic/elastic-agent/commits/2b60ccf8da06fca193cb37a1186c2baa5c3b1f28
Ubuntu 22.04 LTS

Works as expected, is there any os-specific variant that requires further testing ? For example is it required to also test on `Amazon Linux 2` ?

axw commented:
@SylvainJuge thanks! To the best of my knowledge there's no need to test on any specific Linux distros. If anyone knows otherwise, it would be @eyalkoren 

I'll go ahead and mark this tested then.

SylvainJuge commented:
FYI @eyalkoren is on PTO, but I remember reproducing the bug before this change was applied, thus if it used to not work at all and now at least works on one of Linux distributions, so we already have some positive progress.
